# a4 - Data Analysis
Fill in the below code cells as specified. Note that cells may utilize variables and functions defined in previous cells; we should be able to use the `Kernal > Restart & Clear All` menu item followed by `Cell > Run All` to execute your entire notebook and see the correct output.

## Part 1. Numbers
For this part of the assignment, you will analyze some numeric data (counts of library holdings) to investiate how the distribution of numbers in natural data sets obeys the counter-intuitive [Benford's Law](https://plus.maths.org/content/os/issue9/features/benford/index). 

<small>(This exercise was adapted from Steve Wolfman).</small>

Create a variable **`holdings_data`** which is a **list** of the contents of the **`data/libraryholdings.txt`** file included in the repository (each line in the file should be a single element in the list). You will need to open up the file and read its contents into a list. You should specify a _local path_ to the file (from this notebook's location).

In [1]:
holdings_data = []
with open('data/libraryholdings.txt') as file:
    for line in file:
        holdings_data.append(line)

Print out the first **ten** items from the `holdings_data` list, each on its own line. (Note that there may be extra line breaks that are included in the data items themselves).

In [2]:
print(len(holdings_data))
holdings_data[:10]


9142


['(* Library holdings (# of books in each library), *)\n',
 '(* collected by Christian Ayotte.                 *)\n',
 '(* Labels not available.                          *)\n',
 '\n',
 '12201\n',
 '600778\n',
 '14926\n',
 '37863\n',
 '14866\n',
 '9896\n']

Use the **slice operator (`:`)** to remove the "heading" and blank elements from the beginning of the data list, leaving only the list of numbers. The remaining values should continue to be stored (re-stored) in the `holdings_data` variable. Output the new first element in `holdings_data` to demonstrate that it is the first number in the data set.
- The values in the list _should_ be strings rather than an integers

In [3]:
holdings_data = holdings_data[4:]
print(len(holdings_data))

9138


In [4]:
print(holdings_data[0])

12201



Create a variable **`lead_digit_counts`** that is a dictionary whose keys are _strings_ of each digit (`"0"`, `"1"`, `"2"`, etc.), and whose values are all the number `0`. You can do this directly or with a loop. Print out the variable after you create it.

In [5]:
lead_digit_counts={}
for i in range(0,10):
    lead_digit_counts[str(i)]=0
lead_digit_counts

{'0': 0,
 '1': 0,
 '2': 0,
 '3': 0,
 '4': 0,
 '5': 0,
 '6': 0,
 '7': 0,
 '8': 0,
 '9': 0}

Calculate the number of times each digit appears as the _first digit_ in a value of the `holdings_data` list, storing those counts in the `lead_digit_counts` dictionary.

In [6]:
for i in range(0,len(holdings_data)):
    for j in range(0,10):
        if(holdings_data[i][0]==str(j)):
            lead_digit_counts[str(j)]=lead_digit_counts[str(j)]+1

In [7]:
print(lead_digit_counts)

{'0': 0, '1': 3056, '2': 1606, '3': 1018, '4': 801, '5': 640, '6': 560, '7': 502, '8': 503, '9': 452}


Use a loop to print out each count in `lead_digit_counts` with the format:
```
X values have a leading digit of digit Y
```

In [8]:
for key,value in lead_digit_counts.items():
    print(value, 'values have a leading digit of digit', key)

0 values have a leading digit of digit 0
3056 values have a leading digit of digit 1
1606 values have a leading digit of digit 2
1018 values have a leading digit of digit 3
801 values have a leading digit of digit 4
640 values have a leading digit of digit 5
560 values have a leading digit of digit 6
502 values have a leading digit of digit 7
503 values have a leading digit of digit 8
452 values have a leading digit of digit 9


Print the _percentage_ of values in the the library holdings data set that have a leading digit **`1`** (round to 2 decimal places). Is this value as predicted by Benford's law?

In [9]:
n = lead_digit_counts['1']/len(holdings_data)
round(n*100,2)

33.44

Yes it is approximately equal to the value predicted by Benfords law which is 30%. (Source - Wikipedia)

***Extra credit challenge:*** Create a single variable `digit_position_counts` that contains the number of times that each digit 0 through 9 appears in _each_ position in the data set. E.g., a `1` appears in the 1st position 3056 times and in the second position 1005 times; a `2` appears in the 1st position 1606 times and in the second position 1044 times.

Use this variable to print a "table" of the percentage of the time each position contains each digit (e.g., the 1st digit is a `1` 33.44% of the time, a `2` 17.57% of the time, etc).

Note that for this extra challenge it is up to you to determine an appropriate data structure (e.g., how to combine dictionaries and lists and tuples) for representing this table. Be sure and include comments explaining your work.

Only attempt this problem once you have completed everything else!

In [10]:
# Creating a list of dictionaries, where each dictionary has position counts for each digit
list_of_dicts = []
for x in range(0,10):
    digit_count={'0':0,'1':0,'2':0,'3':0,'4':0,'5':0,'6':0,'7':0,'8':0,'9':0}
    for i in range(0,len(holdings_data)):
        for j in range(0,10):
            if(len(holdings_data[i])>x):
                if(holdings_data[i][x]==str(j)):
                    digit_count[str(j)]=digit_count[str(j)] + 1
    list_of_dicts.append(digit_count)

# Using pandas to create a table which has coloumns as digits and rows as positions
import pandas as pd
digit_position_counts = pd.DataFrame(list_of_dicts,index=range(1,11))
print(digit_position_counts)

#Calculating percentages for each element in the dataframe
b = []
for i in range(0,10):
    b.append(digit_position_counts.iloc[i].sum())
len(b)
c = digit_position_counts.divide(b,axis=0)


       0     1     2     3    4    5    6    7    8    9
1      0  3056  1606  1018  801  640  560  502  503  452
2   1162  1005  1044   971  879  889  859  795  762  772
3   1150   890   890   901  866  930  917  868  903  820
4   1299   837   885   844  885  910  782  927  846  906
5   1052   660   713   687  720  820  762  738  719  742
6    161   140   140   153  137  150  126  166  144  164
7     10    14     8     7    5    8   15    9   11    6
8      0     0     0     0    0    0    0    0    0    1
9      0     0     0     0    0    0    0    0    0    0
10     0     0     0     0    0    0    0    0    0    0


In [11]:
#multiply each element by hundred
table = c.multiply(100,axis=0)
table

0          1          2          3         4          5  \
1    0.000000  33.442766  17.574962  11.140293  8.765594   7.003721   
2   12.716130  10.998030  11.424819  10.625958  9.619173   9.728606   
3   12.588944   9.742748   9.742748   9.863164  9.480022  10.180624   
4   14.241859   9.176625   9.702883   9.253371  9.702883   9.976976   
5   13.818468   8.669381   9.365559   9.024038  9.457507  10.771050   
6   10.871033   9.453072   9.453072  10.330858  9.250506  10.128292   
7   10.752688  15.053763   8.602151   7.526882  5.376344   8.602151   
8    0.000000   0.000000   0.000000   0.000000  0.000000   0.000000   
9         NaN        NaN        NaN        NaN       NaN        NaN   
10        NaN        NaN        NaN        NaN       NaN        NaN   

            6          7          8           9  
1    6.128256   5.493543   5.504487    4.946378  
2    9.400306   8.699934   8.338805    8.448238  
3   10.038314   9.501916   9.885057    8.976464  
4    8.573621  10.163359   9.275299    9.933121  
5   10.009195   9.693945   9.444371    9.746486  
6    8.507765  11.208643   9.723160   11.073599  
7   16.129032   9.677419  11.827957    6.451613  
8    0.000000   0.000000   0.000000  100.000000  
9         NaN        NaN        NaN         NaN  
10        NaN        NaN        NaN         NaN

## Part 2. Life Expectancy
For this part of the assignment, you'll work with data about the life expectancy (in years) for each country in the world in the years 1960 and 2013. Note that this can be really [fun](http://www.ted.com/talks/hans_rosling_shows_the_best_stats_you_ve_ever_seen.html) data!

The data is found in a [.csv](https://en.wikipedia.org/wiki/Comma-separated_values) file: a plain-text data format where each line represents a record (row) of data and where feature (column) is separated by a comma.

Read in the contents of the **`data/life_expectancy.csv`** data file, and use it to construct a **list** called **`life_expectancy_list`**. Each element in this list should be a **dictionary** (one for each row in the `csv` file) with the following keys and values:

- a key `'country'` whose value is the name of the country (as a string)
- a key `'le_1960'` whose value is the life expectancy in 1960 (as a float)
- a key `'le_2013'` whose value is the life expectancy in 2013 (as a float)

Thus the first record should look like:
```
{'country': 'Aruba', 'le_1960': 65.56936585, 'le_2013': 75.33217073}
```

You should use the **`csv`** module to read this file and break up each row into different values. See [the documentation](https://docs.python.org/3/library/csv.html) for an example of how to do this. Print out the _first row_ of your list as a demonstration that you've processed the data correctly.

In [12]:
import csv
life_expectancy_list=[]
with open('data/life_expectancy.csv') as csvfile:
    file = csv.reader(csvfile)
    for row in file:
            dictionary={}
            dictionary['country']=row[0]
            dictionary['le_1960']=row[-2]
            dictionary['le_2013']=row[-1]
            life_expectancy_list.append(dictionary)
life_expectancy_list = life_expectancy_list[1:]            
print(life_expectancy_list[0])

{'country': 'Aruba', 'le_1960': '65.56936585', 'le_2013': '75.33217073'}


In [13]:
type(life_expectancy_list[0])

dict

Add another item to each dictionary in the `life_expectancy_list` whose **key** is `change` and whose **value** is the change in life expectancy from 1960 to 2013.

In [14]:
for i in life_expectancy_list:
    a = float(i['le_2013']) - float(i['le_1960'])
    i.update({'change':a})        
#print(life_expectancy_list)

Create a variable **`num_small_gain`** that stores the **number of countries** whose life expectancy did not improve by 5 years or more between 1960 and 2013. This will include counties whose life expectancy has worsened. Print out this variable.

In [15]:
a = []
for i in life_expectancy_list:
    if(i['change']<5):
        a.append(i)

In [16]:
num_small_gain = len(a)
print(num_small_gain)

7


Create a variable **`most_improved`** that is the **name of the country** with the largest gain in life expectancy (between 1960 and 2013). Print out this variable.

In [17]:
b=[]
for i in life_expectancy_list:
    b.append(i['change'])
max_val = max(b)
print(max_val)

for i in life_expectancy_list:
    if(i['change']==max_val):
            a = i['country']
most_improved = a
print(most_improved)

42.07575609
Maldives


Define a function **`compare_country_le()`** that takes in the names of _two_ countries, and returns a **tuple** containing the following information:
- the name of the country with the greater life expectancy,
- the life expectancy in 2013 of that country
- the difference between the life expectancies in 2013

Use your function to print the comparison between the life expectancies of the _United States_ and _Cuba_.  

In [18]:
def compare_country_le(first,second):
    """
    Takes in the names of two countries, and returns a tuple containing the following information:
    
    -the name of the country with the greater life expectancy,
    -the life expectancy in 2013 of that country
    -the difference between the life expectancies in 2013
    
    """
    for i in life_expectancy_list:
        if(i['country']==first):
            first_country = i
        elif(i['country']==second):
            second_country = i
    if(first_country['le_2013'] > second_country['le_2013']):
        name=first_country['country']
        life_exp = first_country['le_2013']
    else:
        name = second_country['country']
        life_exp = second_country['le_2013']
    diff = abs(float(first_country['le_2013'])-float(second_country['le_2013']))
    return (name,life_exp,diff)    

In [19]:
compare_country_le('United States','Cuba')

('Cuba', '79.23926829', 0.39780487999999536)

## Part 3. Readability
For this part of the assignment, you will calculate the [readability](https://en.wikipedia.org/wiki/Readability) of a text document using the [Dale-Chall Readability Formula](http://www.readabilityformulas.com/new-dale-chall-readability-formula.php). This method determines how "easy" it is to read a particular (English) document by considering the length of sentences and how many of the words used are "easy" to understand (based on a pre-defined list of "easy" words).

Splitting real-world text documents into words and sentences is non-trivial (English is hard!). To make this easier, you should use the [Natural Language Toolkit (nltk)](http://www.nltk.org/index.html) module. This module is included with Anacaonda, but does require some additional data source files to be installed on your computer. You _should_ be able to do this by running the below cell (you only need to run it once).

In [20]:
from nltk import download
download('punkt')
download('wordnet')

[nltk_data] Downloading package punkt to
[nltk_data]     /Users/sahilaggarwal/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/sahilaggarwal/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

You will also need to load the list of "easy" words into memory. This list can be found in the **`data/dale-chall.txt`** file. Open this file and read its entire contents into a **list** variable (e.g., `easy_words_list`), where each element in the list is a single line (word) in the file.

In [21]:
with open("data/dale-chall.txt") as file:
    easy_words_list = []
    for line in file:
        easy_words_list.append(line.strip())

In order to "look up" easy words, convert the easy words list into a **dictionary** (e.g., `easy_words_dict`), where each **key** is a word, and each **value** is `True` (that the word is in the list).
- Make sure you do not include newline characters in your keys!

In [22]:
easy_words_dict={}
for i in easy_words_list:
    easy_words_dict[i]=True
#print(easy_words_dict)

Additionally, define a dictionary **`readability_grade_dict`** to use for looking up the "grade level" associated with a readability score (see [this table](https://en.wikipedia.org/wiki/Dale%E2%80%93Chall_readability_formula#Formula)). This dictionary should have **keys** that are ___tuples___ giving the range of score for a particular grade (e.g., `(5.0, 5.9)`), and **values** that are ___strings___ representing the grade (e.g., `"5th or 6th grade"`). 

In [23]:
readability_grade_dict={}
readability_grade_dict[(0.0,4.9)]="4th grade or lower"
readability_grade_dict[(5.0,5.9)]="5th or 6th grade"
readability_grade_dict[(6.0,6.9)]="7th or 8th grade"
readability_grade_dict[(7.0,7.9)]="9th or 10th grade"
readability_grade_dict[(8.0,8.9)]="11th or 12th grade"
readability_grade_dict[(9.0,9.9)]="13th or 15th grade"

Define a function **`print_grade()`** that takes in a readability score (a number greater than or equal to 0), and **prints** a string representing the grade associated with that score (from your `readability_grade_dict` dictionary).
- _Hint:_ loop through the items in the dictionary and determine which "tuple" key has elements that the score falls between. Be sure and round to the nearest decimal).

In [24]:
def print_grade(readibility_score):
    """
    Takes in a readability score (a number greater than or equal to 0), and prints a string representing the grade associated with that score
    """
    grade=0
    if 0.0 < readibility_score < 9.9:
        for i in readability_grade_dict.keys():
            if i[0] < readibility_score < i[1]:
                grade = readability_grade_dict[i]
    else:
        print("Invalid readability score")
    return grade

In [25]:
print_grade(2.2)

'4th grade or lower'

Now to calculate the readability scores! Define a function **`count_sentences()`** that counts the number of sentences in a string. Use the [sent_tokenize()](http://www.nltk.org/api/nltk.tokenize.html#nltk.tokenize.sent_tokenize) function from the `nltk.tokenize` module to break up a string into sentences (this is like the string `split()` function, but it splits into sentences rather than dividing by spaces).
- For help and an example, see [this guide](http://textminingonline.com/dive-into-nltk-part-ii-sentence-tokenize-and-word-tokenize).
- You do not need to do any extra processing beyond that provided by the `sent_tokenize()` function.
- Test your function on a simple pair or trio of sentences!

In [26]:
from nltk.tokenize import sent_tokenize

def count_sentences(text):
    """
    Counts the number of sentences in a string - text
    """
    sentences = sent_tokenize(text)
    number_of_sentences = len(sentences)
    return number_of_sentences

Define a function **`extract_words()`** that takes in a string and returns a _list_ of all of the words in that string. Use the [word_tokenize()](http://www.nltk.org/api/nltk.tokenize.html#nltk.tokenize.word_tokenize) function from the `nltk.tokenize` module to break up the string into words.
- The `nltk` tokenizer includes each punctuation character (e.g., commas, periods) as individual "words". Your list should not include these items. You can use a string method to determine whether or not the word starts with a punctuation symbol, and if so exclude it. _Hint_ think about keeping good words, rather than throwing away the bad! Note that you do not need to do any special consideration for contractions or other words that include their own punctuation.
- Test your function on a simple sentence (with punctuation!).

In [27]:
from nltk.tokenize import word_tokenize
import string
def extract_words(string1):
    """
    Takes in a string (string 1) and returns a list of all of the words in that string
    """ 
    list_words = word_tokenize(string1)
    final_words=[]
    for word in list_words:
        if word[0].isalpha():
            final_words.append(word.lower())
    return final_words

In [28]:
extract_words('Damn!''- Sahil is @@@@@@@ Amazing')

['damn', 'sahil', 'is', 'amazing']

Define a function **`count_easy_words()`** that takes in a _list_ of words as an argument and returns the number of words that are "easy".

- Your function should look up each word in the `easy_words_dict` you defined earlier. _Do not look up words in the list_ (the dictionary is much faster!). Be careful to look up lowercase versions of the word.

- Your function should handle detecting different parts of speech (e.g., plurals, different verb conjugations, etc.). You can do this by using the **`WordNetLemmatizer()`** function from the `nltk.stem.wordnet` module&mdash;which produces a "lemmatizer" object. You can call the **`lemmatize()`** method on this object to reduce a word to its "base" form. See [this example](https://pythonprogramming.net/lemmatizing-nltk-tutorial/). Note that you should reduce words to both their basic noun AND verb forms (you will need to call the function twice: once with `'n'` (noun) and once with `'v'` (verb) as the second argument!)

- You can test your function on the word list: `['My','words','spoken','have','consequences']`, which should have 4 of the 5 words considered easy (not "consequences").

In [29]:
from nltk.stem import WordNetLemmatizer

In [30]:
def count_easy_words(words):
    """
    Takes in a list of words (words) as an argument and returns the number of words that are "easy" comparing it with the 
    easy_words_dict defined earlier.
    """    
    x=[]
    lemmatizer = WordNetLemmatizer()
    words=map(lambda x:x.lower(),words)
    for i in words:
        a=lemmatizer.lemmatize(i,pos='v')
        b=lemmatizer.lemmatize(i,pos='n')
        if(a in easy_words_dict.keys()):
            x.append(i)
        elif(b in easy_words_dict.keys()):
            x.append(i)
    return len(x)

In [31]:
count_easy_words(['My','words','spoken','have','consequences'])

4

Define a function **`calc_readability_score()`** that takes in a string of text and returns a readability "score" for the test based on the [Dale-Chall readability formula](https://en.wikipedia.org/wiki/Dale%E2%80%93Chall_readability_formula#Formula). Call your previous functions to calculate the number of sentences, total words, and number of difficult (not easy) words.
- Don't forget to adjust the score if the text is more than 5% difficult words!

In [32]:
def calc_readability_score(text):
    """
    Takes in a string of text and returns a readability "score" for the test based on the Dale-Chall readability formula.
    """
    const_1 = 0.1579
    const_2 = 0.0496
    no_of_sentences = count_sentences(text)
    words = extract_words(text)
    no_of_words=len(words)
    no_of_easy_words = count_easy_words(words)
    no_of_diff_words = no_of_words-no_of_easy_words
    perc_diff = (no_of_diff_words/no_of_words)*100
    perc_words = (no_of_words/no_of_sentences)
    score = (perc_diff*const_1) + (perc_words*const_2)
    if (perc_diff > 5 ):
        score+=3.6365
    print(no_of_sentences)
    print(no_of_words)
    print(no_of_diff_words)
    return score

Read in the text of the `data/alice.txt` file (the full text of Alice in Wonderland) _as a single string_. 

In [33]:
with open('data/alice.txt','r') as file:
    data = file.read()

In [34]:
#print (extract_words(data))

Calculate the readability score for the `alice.txt` file and print it out. Then print out the reading grade associated with that score. Use your previously-defined functions!
- For testing, note that my calculations show `alice.txt` has 977 sentences and 27198 words, of which 3610 are difficult. This leads to a readability score of ~7.113.

In [35]:
grade = calc_readability_score(data)
print(grade)

977
27198
3610
7.113090902410715


In [36]:
print_grade(grade)

'9th or 10th grade'

_Note that this result may not be an especially accurate model of a text's readability&mdash;after all, it's just based on a simple estimation!_